### PR TITLE
Fix resolve root IServiceProvider

### LINKED ISSUE #36060: The HostBuilder holds on to Autofac's root IServiceProvider which cannot be disposed
### Describe the bug
This is a follow up on https://github.com/autofac/Autofac.Extensions.DependencyInjection/issues/68

The `HostBuilder` creates the initial `IServiceProvider _appServices` for the host:
https://github.com/dotnet/extensions/blob/v3.1.1/src/Hosting/Hosting/src/HostBuilder.cs#L233
For the Autofac integration, this service provider holds a reference to the built Container and certain features of Autofac expect proper disposal of this Container, for example
- instances handed over to Autofac with `builder.RegisterInstance([IDisposable])` live in the container until used by other components
- All active child lifetime scopes are disposed, which is relevant for the `IConfiguration` infrastructure

### To Reproduce
https://github.com/Tragetaschen/autofac-does-not-dispose

There is a `MyDisposable` instance registered with Autofac that is not disposed when the program exists gracefully.

### Expected behavior

My expectation is that I can properly dispose the resources held by `HostBuilder`, but currently there isn't an interface for that. I have been able to restore Autofac's documented behavior by pulling out the private `_appServices` field using reflection and calling Dispose on that.

### Additional context
.NET Core 3.1.1

### PR DESCRIPTION
**Pull request by galakt**
Fixes #36060

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs [hunk 1]
```diff
@@ -214,7 +216,14 @@ private void CreateServiceProvider()
 #pragma warning restore CS0618 // Type or member is obsolete
             services.AddSingleton<IHostApplicationLifetime, ApplicationLifetime>();
             services.AddSingleton<IHostLifetime, ConsoleLifetime>();
-            services.AddSingleton<IHost, Internal.Host>();
+            services.AddSingleton<IHost>(serviceProvider=>
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:219]: nit: spacing

```suggestion
            services.AddSingleton<IHost>(serviceProvider =>
```

**Review by maryamariyan — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs [hunk 2]
```diff
@@ -214,7 +216,14 @@ private void CreateServiceProvider()
...
+            {
+                return new Internal.Host(_appServices,
+                    serviceProvider.GetRequiredService<IHostApplicationLifetime>(),
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:222]: would serviceProvider and _appServices be the same then?

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:222]: Yes I don't like that this code is using different service providers.

File: src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs (hunk 1 above)
Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:219]: Would be cleaner to change this to:

```c#
            services.AddSingleton<IHost>(_ =>
            {
                return new Internal.Host(_appServices,
                    _appServices.GetRequiredService<IHostApplicationLifetime>(),
                    _appServices.GetRequiredService<ILogger<Internal.Host>>(),
                    _appServices.GetRequiredService<IHostLifetime>(),
                    _appServices.GetRequiredService<IOptions<HostOptions>>());
            });
```

File: src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs (hunk 2 above)
Comment (galakt) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:222]: At this lines serviceProvider (ServiceProviderEngineScope) != _appServices (ServiceProvider).
The point of PR that `Host` will have ref to the same root IServiceProvider (_appServices) as in `HostBuilder`

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**